sycl/test-e2e/README.md

```diff
@@ -369,3 +369,12 @@ temporarily disabled due to some issue.
 If you add `UNSUPPORTED` without `UNSUPPORTED-TRACKER` or `UNSUPPORTED-INTENDED`
 directive, the `no-unsupported-without-tracker.cpp` test will fail, notifying
 you about that.
+
+To disable the test completely, you can use:
+```
+// USNUPPORTED: overall *OR* hangs *OR* flaky
+```
+See `unsupported_feature` set in sycl/test-e2e/lit.cfg.py for these features.
+
+Note: please do not use `REQUIRES: TEMPORARY_DISABLED`. Use `UNSUPPORTED`
```

Comment:
Please, add clarification, why and when `REQUIRES` should be avoided. LLVM's testing guide has following tips: https://llvm.org/docs/TestingGuide.html#tips-for-writing-constraints.

Reply:
> Generally, people use `REQUIRES` to state things that the test depends on to operate correctly, and `UNSUPPORTED` to exclude cases where the test is expected never to work.

`REQUIRES: TEMPORARY_DISABLED` is explicitly the case where the test is expected never to work, so `UNSUPPORTED` should be used.

But our main motivation is to switch to a mechanism which we track: `REQUIRES: TEMPORARY_DISABLED` can easily go without any trackers being created, leaving that test disabled **for years**:
- `sycl/test-e2e/Basic/image/image_accessor_range.cpp` - has been disabled for 2 years
- `sycl/test-e2e/Regression/complex_global_object.cpp` - has been disabled for at least 4 years

We have a mechanism to track that every `UNSUPPORTED` directive either has an explanation about why is it intended, or has a tracker associated with it, see https://github.com/intel/llvm/blob/sycl/sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp

Reply:
I'm not sure I understand the problem with tracking tests with `REQUIRES` instead of `UNSUPPORTED`. How are you going to control which keyword is used? Don't we replace one problem with another?

Reply:
I'm not convinced that we have to create artificial limitations for developers to enable the tracking. I believe that we can catch misuses of `REQUIRES` keyword using the options I mentioned in [this comment](https://github.com/intel/llvm/pull/15946#issuecomment-2455482606).

Reply:
Just interesting - I plan to add a test to check if all features passed to `REQUIRES` exist. Will it be then a limitation to prohibit developers to put there everything they want?

Reply:
Asking because TEMPORARY_DISABLED is not documented anywhere (or at least I didn't see it)

Reply:
> I'm not convinced that we have to create artificial limitations for developers to enable the tracking. I believe that we can catch misuses of `REQUIRES` keyword using the options I mentioned in [this comment](https://github.com/intel/llvm/pull/15946#issuecomment-2455482606).

So, we do use them (at least some of them) in pre-commit and I do not see how it helps. Having a developer to go look into their pre-commit and make sure that their new/modified test are not listed as skipped/unsupported is even less reliable than code review.

Moreover, to detect tests which are **already** accidentally skipped, we would have to review them all manually. Or write a script I was talking about, which makes us one step away from enabling it as a test.

Considering that the situation when a test is completely disabled everywhere is relatively rare (we only have ~10 tests like this in the whole 1000+ tests codebase), I do not think that any limitations we are imposing are severe.

Reply:
> Just interesting - I plan to add a test to check if all features passed to `REQUIRES` exist. Will it be then a limitation to prohibit developers to put there everything they want?

I think this is want we want.

Reply:
Just to try and converge the discussion here, the intention of this line in the README is to consolidate the ways we disable tests. Previously, we would have `UNSUPPORTED` in some cases, `XFAIL` in others, and `REQUIRES: TEMPORARILY_DISABLED` in a few. Now we want to unify these, which we do by:

1. Using `UNSUPPORTED` for cases that are intentionally unsupported for a given feature, e.g. using Intel device only features and disabling for CUDA and HIP. Note that `REQUIRES` is also an option for these, but sometimes exclusion is better than inclusion.
2. Using `UNSUPPORTED` for cases where the tests are currently unintentionally flakily failing or hangs happen.
3. Using `XFAIL` for cases where tests are unintentionally failing, excluding the cases from the previous point.

That leaves `REQUIRES` to be used exclusively for intentional requirements. From a tracking and code-quality POV, I argue this comes with the following benefits:

- Developers are required to add JIRA and/or Github trackers when disabling tests, which helps us make sure that these failures don't go uninvestigated for a _long_ time, as we have seen happen before.
- Gathering statistics on unintentionally failing tests is a lot easier.
- It allows us to make a script that can check for known features in `REQUIRES` fields to avoid cases where a misspelling of a feature unintentionally disables a test on all targets. Technically, we could include `TEMPORARILY_DISABLED` as a "known" feature, but I see no point in doing so when we have alternatives that should be used instead, given the two points above.

@bader - Do I understand correctly that the initial point of this comment thread was to refine the wording here? If so, would it be something like:

> Please do not use `REQUIRES` to temporarily disable a given test on targets that unintentionally fail it. Use `XFAIL` or `UNSUPPORTED` instead.

Reply:
@bader do you still think we shouldn't merge this one?

Reply:
> @bader - Do I understand correctly that the initial point of this comment thread was to refine the wording here?

Yes.

> If so, would it be something like:
> 
> > Please do not use `REQUIRES` to temporarily disable a given test on targets that unintentionally fail it. Use `XFAIL` or `UNSUPPORTED` instead.

This wording addresses only `when` part of my request, but not `why`. Based on the answers in this thread, it sounds like the reason to avoid `REQUIRES` is that it's easier to misuse then `UNSUPPORTED`. Did I get it right?

Reply:
> @bader do you still think we shouldn't merge this one?

I don't object to merging, but I have doubts that this PR eliminates the root cause of the problem with today's using `REQUIRES`. I believe there are better ways to track if there are misuses of llvm-lit keywords.

@KornevNikita, please, resolve merge conflicts before merging.

Reply:
> Based on the answers in this thread, it sounds like the reason to avoid REQUIRES is that it's easier to misuse then UNSUPPORTED. Did I get it right?

As one of the reasons. I'd also say that `TEMPORARY_DISABLED` is some kind of work-around. And we have a more official alternative in the form of `UNSUPPORTED`, which we are also already tracking.

Reply:
> > Based on the answers in this thread, it sounds like the reason to avoid REQUIRES is that it's easier to misuse then UNSUPPORTED. Did I get it right?
> 
> As one of the reasons. I'd also say that `TEMPORARY_DISABLED` is some kind of work-around. And we have a more official alternative in the form of `UNSUPPORTED`, which we are also already tracking.

I don't agree with that statement. Both are viable options. I guess you call the use of `UNSUPPORTED` keyword "official" because it's documented. We could make use of `REQUIRES` official as well if we wanted. 

I'm okay with the idea of setting the convention to use `UNSUPPORTED` for SYCL tests disabling.